components/centraldashboard/public/components/resource-chart.js

```diff
@@ -0,0 +1,352 @@
+        return {
+            headerText: String,
+            metric: String,
+            interval: String,
+            externalLink: String,
+            externalLinkText: String,
+            displayedHeader: {
+                type: String,
+                computed: '_getDisplayedHeader(headerText, truncatedSeries)',
+            },
+            metricUrl: {
+                type: String,
+                computed: '_getUrl(metric, interval)',
+            },
+            truncatedSeries: {
+                type: Boolean,
+                readOnly: true,
+                value: false,
+            },
+        };
+    }
+
+    constructor() {
+        super();
+        this._chartOptions = {
+            type: 'line',
+            data: {
+                datasets: [],
+            },
+            options: {
+                maintainAspectRatio: false,
+                tooltips: {
+                    mode: 'index',
+                    intersect: false,
+                    callbacks: {
+                        label: this._buildTooltipsLabel.bind(this),
+                    },
+                },
+                scales: {
+                    xAxes: [{
+                        type: 'time',
+                        gridLines: {
+                            display: false,
+                        },
+                        distribution: 'series',
+                        ticks: {
+                            source: 'auto',
+                        },
+                    }],
+                    yAxes: [{
+                        type: 'linear',
+                    }],
+                },
+                legend: {
+                    display: true,
+                    position: 'bottom',
+                    labels: {
+                        boxWidth: 6,
+                        usePointStyle: true,
+                    },
+                },
+                plugins: {
+                    crosshair: {
+                        line: {
+                            color: '#dadce0',
+                            width: 1,
+                        },
+                        zoom: {
+                            enabled: false,
+                        },
+                        sync: {
+                            enabled: false,
+                        },
+                    },
+                },
+            },
+        };
+    }
+
+    /**
+     * Instantiates the Chart object using the canvas context.
+     */
+    ready() {
+        super.ready();
+        this._chart = new Chart(this.$.chart, this._chartOptions);
+    }
+
+    _refresh() {
+        this.$.ajax.generateRequest();
+    }
+
+    /**
+     * Handles the Metrics response to build datasets and create the Chart.
+     * @param {Event} responseEvent
+     */
+    _onResponse(responseEvent) {
+        let i = 0;
+        this._chartOptions.data.datasets = [];
+        for (const ds of this._buildDatasets(responseEvent.detail.response)) {
```

Comment:
#### suggestion:
Remove lines 230-231
```suggestion
        const dataSets = this._buildDatasets(responseEvent.detail.response)
        const chartSets = dataSets.map((ds, i) => ({
            /* direct object creation and return */
        }))
        this._chartOptions.data.datasets = chartSets
```

Reply:
Great suggestion. I sometimes forgot that the functional methods for JS arrays are more powerful and accept additional callback parameters besides the element. I also adopted your style for the implicit return for a single object in a lambda.